src/corppa/poetry_detection/annotation/recipe.py

```diff
@@ -0,0 +1,172 @@
+"""
+This module provides custom recipes for Prodigy annotation. They were
+created with page-level text annotation in mind, and support annotating
+text with a reference image displayed beside the text (`annotate_page_text`),
+or annotating both text and image side by side (`annotate_text_and_image`).
+
+Referenced images must be served out independently for display.
+
+Example use:
+```
+prodigy annotate_page_text poetry_spans poetry_pages.jsonl --label POETRY,PROSODY -F ../corppa/poetry_detection/annotation/recipe.py --image-prefix http://localhost:8000/
+prodigy annotate_text_and_image poetry_text_image poetry_pages.jsonl --label POETRY,PROSODY -F ../corppa/poetry_detection/annotation/recipe.py --image-prefix http://localhost:8000/
+"""
+
+from prodigy.core import Arg, recipe
+from prodigy.components.loaders import JSONL
+import spacy
+
+from pathlib import Path
+
+#: reference to current directory, for use as Prodigy CSS directory
+CURRENT_DIR = Path(__file__).parent.absolute()
+
+#: common prodigy configurations for both recipes; copy and add blocks and labels
+PRODIGY_COMMON_CONFIG = {
+    "show_flag": True,  # show flag button to mark weird/difficult examples
+    "hide_newlines": False,  # ensure newlines are shown \n
+    "allow_newline_highlight": True,  # allow highlighting of newlines \n
+    "honor_token_whitespace": True,  # reflect whitespace accurately (e.g. in case of leading/trailing spaces)
+    "custom_theme": {
+        "labels": {
+            # trying to use options from PPA webapp color scheme,
+            # but may not be so great in Prodigy UI.
+            # azure #0788fc seafoam blue #57c4c4 wisteria #9c93c0 pig pink #ed949c
+            "POETRY": "#57c4c4",  # label color for POETRY
+        },
+        "hide_true_newline_tokens": False,
+    },
+    "global_css_dir": CURRENT_DIR,
+}
+
+
+def tokenize_stream(stream, image_prefix=None):
+    """Takes a stream of Prodigy tasks and tokenizes text for span annotation,
+    and optionally adds an image prefix to any image path present.
+    Stream is expected to contain `text` and may contain image_path` and a `meta`
+    dictionary. Returns a generator of the stream.
+    """
+
+    nlp = spacy.blank("en")  # use blank spaCy model for tokenization
+
+    # ensure image prefix does not have a trailing slash
+    if image_prefix is None:
+        image_prefix = ""
+    image_prefix = image_prefix.rstrip("/")
+
+    for task in stream:
+        if task.get("text"):
+            doc = nlp(task["text"])
+            task["tokens"] = [
+                {
+                    "text": token.text,
+                    "start": token.idx,
+                    "end": token.idx + len(token.text),
+                    "id": i,
+                }
+                for i, token in enumerate(doc)
+            ]
```

Comment:
This list comprehension is difficult to read, consider converting into a for loop for readability